Add controller Service Instance test

jasiu001 opened the pull request:
This PR is a 
 - [x] Feature Implementation
 - [ ] Bug Fix
 - [ ] Documentation

**What this PR does / why we need it**:
The PR indroduces controller tests which will replace current integration tests, see #2597
Test refers to the ServiceInstance flow:

- TestCreateServiceInstanceFailsWithNonexistentPlan
- TestCreateServiceInstanceNonExistentClusterServiceBroker
- TestCreateServiceInstanceNonExistentClusterServiceClassOrPlan
- TestCreateServiceInstanceWithInvalidParameters
- TestCreateServiceInstanceWithParameters
- TestCreateServiceInstanceWithProvisionFailure
- TestUpdateServiceInstanceChangePlans
- TestUpdateServiceInstanceChangePlansToNonexistentPlan
- TestUpdateServiceInstanceNewDashboardResponse
- TestUpdateServiceInstanceUpdateParameters

Please leave this checklist in the PR comment so that maintainers can ensure a good PR.

Merge Checklist:
 - [ ] New feature 
   - [ ] Tests
   - [ ] Documentation
 - [ ] SVCat CLI flag
 - [ ] Server Flag for config
   - [ ] Chart changes
   - [ ] removing a flag by marking deprecated and hiding to avoid
         breaking the chart release and existing clients who provide a
         flag that will get an error when they try to update

MHBauer commented:
/ok-to-test

MHBauer commented:
/assign @jberkhahn

mszostok commented:
/test pull-service-catalog-integration

jberkhahn commented:
/retest

jberkhahn commented on pkg/controller/case_test.go:
```diff
@@ -934,6 +958,670 @@ func (ct *controllerTest) v1Now() *metav1.Time {
+
+// WaitForInstanceCondition waits until ServiceInstance `status.conditions` field value is equal to condition in parameters
+// returns error if the time limit has been reached
+func (ct *controllerTest) WaitForInstanceCondition(condition v1beta1.ServiceInstanceCondition) error {
```
Is there a  reason you duplicated this instead of using the one in test/util/util.go?

jberkhahn commented on pkg/controller/case_test.go:
```diff
@@ -934,6 +958,670 @@ func (ct *controllerTest) v1Now() *metav1.Time {
+// WaitForServiceInstanceProcessedGeneration waits until ServiceInstance parameter `Status.ObservedGeneration` is
+// equal or higher than ServiceInstance `generation` value, ServiceInstance is in Ready/True status and
+// ServiceInstance is not in Orphan Mitigation progress
+func (ct *controllerTest) WaitForServiceInstanceProcessedGeneration(generation int64) error {
```
same question as above.

jberkhahn commented on pkg/controller/case_test.go:
```diff
@@ -934,6 +958,670 @@ func (ct *controllerTest) v1Now() *metav1.Time {
+	return err
+}
+
+func isServiceInstanceConditionTrue(instance *v1beta1.ServiceInstance) bool {
```
^^

jberkhahn commented on pkg/controller/case_test.go:
```diff
@@ -934,6 +958,670 @@ func (ct *controllerTest) v1Now() *metav1.Time {
+}
+
+// AssertServiceInstanceHasNoCondition makes sure ServiceInstance is in not specific condition
+func (ct *controllerTest) AssertServiceInstanceHasNoCondition(t *testing.T, cond v1beta1.ServiceInstanceCondition) {
```
would it make sense to move this to the util package?

jberkhahn commented on pkg/controller/case_test.go:
```diff
@@ -934,6 +958,670 @@ func (ct *controllerTest) v1Now() *metav1.Time {
+}
+
+// AssertServiceInstanceOrphanMitigationStatus makes sure ServiceInstance is/or is not in Orphan Mitigation progress
+func (ct *controllerTest) AssertServiceInstanceOrphanMitigationStatus(t *testing.T, state bool) {
```
seems like another thing to stick in the util package

jberkhahn commented on pkg/controller/case_test.go:
```diff
@@ -934,6 +958,670 @@ func (ct *controllerTest) v1Now() *metav1.Time {
+}
+
+// CreateClusterServiceClass creates ClusterServiceClass with default parameters
+func (ct *controllerTest) CreateClusterServiceClass() error {
```
Pretty much all of this might want to be moved into test/util. I could have sworn there was an equivalent for this in there already but I can't find it offhand.

jberkhahn commented on pkg/controller/controller_flow_instance_test.go:
```diff
@@ -162,3 +164,755 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
+
+// TestCreateServiceInstanceFailsWithNonexistentPlan tests creating a ServiceInstance whose ServicePlan
+// does not exist
+func TestCreateServiceInstanceFailsWithNonexistentPlan(t *testing.T) {
```
seems good

jberkhahn commented:
The tests themselves seem good, but I think there's quite a bit of refactoring of the setup stuff that could be done. A lot are dupes of functions we already have in the test/util package.

jasiu001 commented:
I noticed that all comments are about why I don't use existing functions in `test/util/util.go`. First of all, we moved all controller integration tests from `test/integration` to `pkg/controller` so it is natural we moved also `util` functions.

As you can see in this [commit](https://github.com/kubernetes-sigs/service-catalog/commit/9f31b7d5f8a9b8804b9223272766a6c60d16cf79) which is already merged to master "utils" functions are moved to `controller` package along with controller's tests.
If I am not mistaken, the `test/integration` will be deleted in the future.

In the end, I would like to point out that in "our" `case_test` we use `controllerTest's` methods instead of functions so we don't have to e.g. inject to each "util" function the client, which in my opinion is much easier to read.

jberkhahn commented:
/approve
@MHBauer could you take a look at this when you get a chance?

MHBauer commented:
Is this primarily movement of existing tests or new tests? I do not really have the time to review 1400 lines.

I cannot tell from the commit message or the PR message if this is new or old tests. Moved or recreated or freshly created.

jasiu001 commented:
@MHBauer this PR is the second part of moved tests (the first part was merged to master, here is the [commit](https://github.com/kubernetes-sigs/service-catalog/commit/9f31b7d5f8a9b8804b9223272766a6c60d16cf79).
As I mentioned above we moved all controller integration tests from `test/integration` to `pkg/controller` and we've rewritten the tests for the new version with CRDs.

mszostok commented:
/retest

mszostok commented:
@MHBauer In general here is the issue about that: 
https://github.com/kubernetes-sigs/service-catalog/issues/2597

To make a long story short, current integration tests are tightly coupled with api-server (starting server, etcd). We want to rewrite them into general integration tests with only one dependency to SC controllers.

Right now, the rewritten tests are placed are under `pkg/controller` package with `// +build integration` tag, so they are executed only if you execute 
```
go test --tags="integration" github.com/kubernetes-sigs/service-catalog/pkg/controller/...
```

> @MHBauer: I'll have to dig into the existing tests to find the corresponding other tests.

the test name is exactly the same as defined in https://github.com/kubernetes-sigs/service-catalog/tree/master/test/integration, so it's easy to check them

e.g. `TestServiceBindingDeleteWithAsyncBindInProgress` 

when you execute search:
https://github.com/kubernetes-sigs/service-catalog/search?q=TestServiceBindingDeleteWithAsyncBindInProgress&unscoped_q=TestServiceBindingDeleteWithAsyncBindInProgress

then you will see that test is defined in both places:
- `test/integration/controller_test.go`
- and `pkg/controller/controller_flow_binding_test.go`

they are testing the same logic but the new way does not require the api-server

right now we are in the transition phase and we run both tests (this is also the reason why `pull-service-catalog-integration` takes a little bit more time), so we can be sure that both working well. After merging CRDs we are able to remove the old integration test (with api-server) but still have the same coverage thanks to new tests which were implemented, so we will not regress.

mszostok commented on pkg/controller/controller_flow_instance_test.go:
```diff
@@ -162,3 +164,755 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
+// TestCreateServiceInstanceNonExistentClusterServiceClassOrPlan tests that a ServiceInstance gets
+// a Failed condition when the service class or service plan it references does not exist.
+func TestCreateServiceInstanceNonExistentClusterServiceClassOrPlan(t *testing.T) {
+	// TODO: cannot rewrite tests because fakeClient does not support `FieldSelector` during filtering ServiceInstance list
```
add info that it should be added after merging the CRDs

jasiu001 replied:
Done.

jasiu001 commented:
@MHBauer considering your comment about parallel tests I paralleled some tests. The list of tests and their execution times before and after are presented in the table below:

test | time execution before | time execution after | notes |
--|--|--|--
TestProvisionInstanceWithRetries | 4.100s |  2.052s |
TestRetryAsyncDeprovision | - | - | single test |
TestServiceInstanceDeleteWithAsyncProvisionInProgress | - | - | already Parallel |
TestServiceInstanceDeleteWithAsyncUpdateInProgress | - | - | already Parallel |
TestCreateServiceInstanceFailsWithNonexistentPlan | - | - | single test |
TestCreateServiceInstanceNonExistentClusterServiceBroker | - | - | single test |
TestCreateServiceInstanceNonExistentClusterServiceClassOrPlan | - | - | - |
TestCreateServiceInstanceWithInvalidParameters | - | - | single test |
TestCreateServiceInstanceWithParameters | 9.288s | 3.104s | |
TestCreateServiceInstanceWithProvisionFailure | 21.232s | 6.048s | |
TestUpdateServiceInstanceChangePlans | 12.521s | 4.266s | |
TestUpdateServiceInstanceChangePlansToNonexistentPlan | - | - | single test |
TestUpdateServiceInstanceNewDashboardResponse | - | - | cannot be run parralel, details in code |
TestUpdateServiceInstanceUpdateParameters | 18.316s | 5.417s | |
**All tests** | 87.120s | 48.008s | |

-----

When it comes to the `TestUpdateServiceInstanceUpdateParameters` test, the problem was related to another test that set the` feature gate` flag to true. The correct performance of the test depended on the order in which the tests were run. The error has been corrected.

-----
I also tried to reproduce failing `TestCreateServiceInstanceWithInvalidParameters` test (with count=100), but each time test passed. I also tried to run this test on @mszostok computer but it passed. Please, I need more details about how do you run this test.

-----
@mszostok missing tests have been added

jasiu001 commented:
/test pull-service-catalog-test-e2e

jasiu001 commented:
/test pull-service-catalog-test-integration

jasiu001 commented:
/test pull-service-catalog-test-e2e

jasiu001 commented:
/test pull-service-catalog-test-e2e

jasiu001 commented:
/test pull-service-catalog-test-e2e

mszostok commented:
/test pull-service-catalog-test-e2e

MHBauer commented:
I forgot about feature gates. I remember those preventing nearly any use of parallel. Oh well.

mszostok commented on pkg/controller/case_test.go:
```diff
@@ -928,12 +953,804 @@ func (ct *controllerTest) WaitForClusterServicePlan() error {
+	}
+}
+
+// TODO: move to case_test.go
```
should be removed, right?

mszostok commented on pkg/controller/case_test.go:
```diff
@@ -928,12 +953,804 @@ func (ct *controllerTest) WaitForClusterServicePlan() error {
+	return err
+}
+
+// TODO: move to case_test.go
```
remove?

mszostok commented on pkg/controller/case_test.go:
```diff
@@ -928,12 +953,804 @@ func (ct *controllerTest) WaitForClusterServicePlan() error {
+	}
+}
+
+// TODO: move to case_test.go
```
^^

mszostok commented on pkg/controller/controller_flow_test.go:
```diff
@@ -59,6 +59,39 @@ func TestBasicFlowWithBasicAuth(t *testing.T) {
 }
 
+// TestOriginatingIdentity tests whether the controller uses correct credentials when the secret changes
+func TestOriginatingIdentity(t *testing.T) {
```
add info
`// CAUTION: the test cannot be executed in parallel because it changes global flag which can affect the behavior of other tests`

MHBauer commented:
Is there a 1second sleep in here somewhere? A lot of tests take that suspiciously specific amount of time.

MHBauer commented:
here is exact invocation:

```console
$ go test --tags="integration"  -v -run TestCreateServiceInstanceWithInvalidParameters -race -count 30 -failfast
=== RUN   TestCreateServiceInstanceWithInvalidParameters
=== PAUSE TestCreateServiceInstanceWithInvalidParameters
=== CONT  TestCreateServiceInstanceWithInvalidParameters
--- PASS: TestCreateServiceInstanceWithInvalidParameters (1.01s)
=== RUN   TestCreateServiceInstanceWithInvalidParameters
=== PAUSE TestCreateServiceInstanceWithInvalidParameters
=== CONT  TestCreateServiceInstanceWithInvalidParameters
--- PASS: TestCreateServiceInstanceWithInvalidParameters (1.01s)
=== RUN   TestCreateServiceInstanceWithInvalidParameters
=== PAUSE TestCreateServiceInstanceWithInvalidParameters
=== CONT  TestCreateServiceInstanceWithInvalidParameters
--- PASS: TestCreateServiceInstanceWithInvalidParameters (1.01s)
=== RUN   TestCreateServiceInstanceWithInvalidParameters
=== PAUSE TestCreateServiceInstanceWithInvalidParameters
=== CONT  TestCreateServiceInstanceWithInvalidParameters
--- FAIL: TestCreateServiceInstanceWithInvalidParameters (8.97s)
    require.go:794: 
        	Error Trace:	controller_flow_instance_test.go:252
        	Error:      	Received unexpected error:
        	            	instance with proper conditions not found, the existing conditions: [{Type:Ready Status:True LastTransitionTime:2019-09-03 14:50:00.909996568 -0700 PDT m=+4.054543438 Reason:ProvisionedSuccessfully Message:The instance was provisioned successfully}]
        	Test:       	TestCreateServiceInstanceWithInvalidParameters
FAIL
exit status 1
FAIL	github.com/kubernetes-sigs/service-catalog/pkg/controller	12.067s
```

Do instances of the same test interfere with each other? Are we not cleaning up after ourselves?

MHBauer commented:
As a separate task, there's some very bizarre imports. Things imported multiple times with different names.

jasiu001 commented:
There is no `sleep` inside any of test, all initiation operation in each test I think to take the same amount of time that is the reason why all have similar time.

As for the test `TestCreateServiceInstanceWithInvalidParameters` we found the reason why sometimes the test failed, unlike the original one in this I create the instance first and then update it with wrong parameters. 
Sometimes informer won't refresh the cache with instances after the update and the test got the instance from before update (`success` instance with no parameters).
I fixed it and now instance is created with wrong parameters at the beginning.

You are right about there were imports with the same library, I found two in `case_test.go` file:
- `github.com/pmorie/go-open-service-broker-client/v2`
- `k8s.io/apimachinery/pkg/apis/meta/v1`

I fixed it.

jasiu001 commented:
/test pull-build-all-images-for-ppc64le

MHBauer commented:
Awesome. I'll give it another soak locally.

MHBauer commented:
There's definitely a goroutine lack-of-cleanup threadsplosion issue, but single runs appear to be okay.

MHBauer commented:
I think, this is
/lgtm

I'll let someone else confirm. Remove the hold when it's ready. 
/hold

I've created #2698 as an issue to look into.

jberkhahn commented:
ship it 
/hold cancel
